experimental/cuda2/pending_queue_actions.c

```diff
@@ -489,26 +644,38 @@ iree_status_t iree_hal_cuda2_pending_queue_actions_issue(
       // If this semaphore has already signaled past the desired value, we can
       // just ignore it.
       uint64_t value = 0;
-      IREE_RETURN_AND_END_ZONE_IF_ERROR(
-          z0, iree_hal_semaphore_query(semaphores[i], &value));
+      iree_status_t status = iree_hal_semaphore_query(semaphores[i], &value);
+      if (!iree_status_is_ok(status)) {
+        iree_slim_mutex_unlock(&actions->action_mutex);
+        IREE_TRACE_ZONE_END(z0);
+        return status;
```

Comment:
lots of early returns here - if possible try to avoid that especially when locks and leakable resources are involved (someone trying to read/modify this code is going to have to veeeryyyy carefully understand what you're doing in order not to introduce bugs)

Reply:
Good point! Changed the code structure to avoid many early returns.